src/crimson/os/seastore/transaction_manager.h

```diff
@@ -857,18 +873,24 @@ class TransactionManager : public ExtentCallbackInterface {
       pref.is_indirect() ?
 	pref.get_intermediate_length() :
 	pref.get_length(),
-      [pin=std::move(pin)]
+      [&pref]
       (T &extent) mutable {
 	assert(!extent.has_laddr());
 	assert(!extent.has_been_invalidated());
-	assert(!pin->has_been_invalidated());
-	assert(pin->get_parent());
-	pin->link_child(&extent);
-	extent.maybe_set_intermediate_laddr(*pin);
+	assert(!pref.has_been_invalidated());
+	assert(pref.get_parent());
+	pref.link_child(&extent);
+	extent.maybe_set_intermediate_laddr(pref);
       }
-    ).si_then([FNAME, &t](auto ref) mutable -> ret {
+    ).si_then([FNAME, &t, pin=std::move(pin)](auto ref) mutable -> ret {
       SUBTRACET(seastore_tm, "got extent -- {}", t, *ref);
       assert(ref->is_fully_loaded());
+      if (pin->is_indirect()) {
+	ceph_assert(pin->get_checksum() == 0);
+      } else {
+	ceph_assert(pin->get_checksum() == 0 ||
+		    pin->get_checksum() == ref->calc_crc32c());
+      }
```

Comment:
After `full_extent_integrity_check` is introduced, is it supposed to be:
```c++
if (pin->is_indirect()) {
  ceph_assert(pin->get_checksum() == 0);
} else if (full_extent_integrity_check) {
  ceph_assert(pin->get_checksum() == ref->calc_crc32c());
} else { // !full_extent_integrity_check: remapped extent may be skipped
  ceph_assert(pin->get_checksum() == 0 ||
              pin->get_checksum() == ref->calc_crc32c());
}
```

Also, checksum mismatch may not be logical bugs -- it may caused by device issues -- so print mismatched information in error log is necessary here, I think.